src/session_manager.rs

```diff
@@ -0,0 +1,989 @@
+                                None
+                            }
+                        })
+                    })
+                    .unwrap_or_default();
+
+                self_
+                    .sessions
+                    .pages()
+                    .select_item(active_session_index, true);
+                // save recently used sessions
+                self.save_recently_used_sessions();
+            } else {
+                // There are no sessions left. Thus go back to login.
+                self.add_new_session(APPLICATION_OPTS.get().unwrap().test_dc);
+            }
+        }
+    }
+
+    /// Function cleaning up, which is called by the application windows on closing.
+    ///
+    /// It sets all clients offline and logs them out. Additionally, if there are logged in
+    /// sessions, it removes all sessions that are in auth state. There can only be one session in
+    /// auth state, btw.
+    pub fn close_clients(&self) {
+        let self_ = imp::SessionManager::from_instance(self);
+
+        if self.sessions().n_items() > 0 {
+            // Create a future to set all logged in session to offline to close them afterwards.
+            let close_sessions_future = futures::future::join_all(
+                self_
+                    .clients
+                    .borrow()
+                    .iter()
+                    .filter_map(|(client_id, client_info)| match client_info {
+                        ClientInfo::LoggedIn(_) => Some(client_id),
+                        _ => None,
+                    })
+                    .cloned()
+                    .map(|client_id| {
+                        set_online(client_id, false)
+                            .and_then(move |_| functions::Close::new().send(client_id))
+                    }),
+            );
+
+            // Create a future to delete all sessions that are currently in auth state.
+            let tear_down_future = futures::future::join_all(
+                self_
+                    .clients
+                    .borrow()
+                    .iter()
+                    .filter_map(|(client_id, client_info)| match client_info {
+                        ClientInfo::Auth { database_info, .. } => {
+                            Some((client_id, &database_info.directory_base_name))
+                        }
+                        _ => None,
+                    })
+                    .map(|(client_id, database_dir_base_name)| {
+                        teardown_database(database_dir_base_name.clone(), *client_id)
+                    }),
+            );
+
+            // Block on that future, else the window closes before they are finished!!!
+            RUNTIME.block_on(async {
+                let (close_sessions, tear_down) =
+                    futures::future::join(close_sessions_future, tear_down_future).await;
+                close_sessions.into_iter().for_each(|r| {
+                    r.unwrap();
```

Comment:
I'm getting a panic here doing the following:

- Accessing with my personal account
- Accessing with a test account
- Closing the app
- Removing the session of my personal account from another device
- Reopening the app and closing it again (also, I don't know if we can log out the client automatically in that case, but that's a different issue)

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { code: 500, message: "Request aborted" }', src/session_manager.rs:564:23
stack backtrace:
   0: rust_begin_unwind
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panicking.rs:100:14
   2: core::result::unwrap_failed
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/result.rs:1616:5
   3: core::result::Result<T,E>::unwrap
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/result.rs:1298:23
   4: telegrand::session_manager::SessionManager::close_clients::{{closure}}::{{closure}}
             at ./Projects/telegrand/src/session_manager.rs:564:21
   5: core::iter::traits::iterator::Iterator::for_each::call::{{closure}}
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/iter/traits/iterator.rs:734:29
   6: core::iter::traits::iterator::Iterator::fold
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/iter/traits/iterator.rs:2171:21
   7: core::iter::traits::iterator::Iterator::for_each
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/iter/traits/iterator.rs:737:9
   8: telegrand::session_manager::SessionManager::close_clients::{{closure}}
             at ./Projects/telegrand/src/session_manager.rs:563:17
   9: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80:19
  10: tokio::park::thread::CachedParkThread::block_on::{{closure}}
             at ./Projects/telegrand/_build/cargo-home/registry/src/github.com-1ecc6299db9ec823/tokio-1.15.0/src/park/thread.rs:263:54
  11: tokio::coop::with_budget::{{closure}}
             at ./Projects/telegrand/_build/cargo-home/registry/src/github.com-1ecc6299db9ec823/tokio-1.15.0/src/coop.rs:102:9
  12: std::thread::local::LocalKey<T>::try_with
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/thread/local.rs:399:16
  13: std::thread::local::LocalKey<T>::with
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/thread/local.rs:375:9
  14: tokio::coop::with_budget
             at ./Projects/telegrand/_build/cargo-home/registry/src/github.com-1ecc6299db9ec823/tokio-1.15.0/src/coop.rs:95:5
  15: tokio::coop::budget
             at ./Projects/telegrand/_build/cargo-home/registry/src/github.com-1ecc6299db9ec823/tokio-1.15.0/src/coop.rs:72:5
  16: tokio::park::thread::CachedParkThread::block_on
             at ./Projects/telegrand/_build/cargo-home/registry/src/github.com-1ecc6299db9ec823/tokio-1.15.0/src/park/thread.rs:263:31
  17: tokio::runtime::enter::Enter::block_on
             at ./Projects/telegrand/_build/cargo-home/registry/src/github.com-1ecc6299db9ec823/tokio-1.15.0/src/runtime/enter.rs:151:13
  18: tokio::runtime::thread_pool::ThreadPool::block_on
             at ./Projects/telegrand/_build/cargo-home/registry/src/github.com-1ecc6299db9ec823/tokio-1.15.0/src/runtime/thread_pool/mod.rs:77:9
  19: tokio::runtime::Runtime::block_on
             at ./Projects/telegrand/_build/cargo-home/registry/src/github.com-1ecc6299db9ec823/tokio-1.15.0/src/runtime/mod.rs:463:43
  20: telegrand::session_manager::SessionManager::close_clients
             at ./Projects/telegrand/src/session_manager.rs:560:13
  21: <telegrand::window::imp::Window as gtk4::subclass::window::WindowImpl>::close_request
             at ./Projects/telegrand/src/window.rs:105:13
  22: gtk4::subclass::window::window_close_request
             at ./Projects/telegrand/_build/cargo-home/registry/src/github.com-1ecc6299db9ec823/gtk4-0.3.1/src/subclass/window.rs:162:5
  23: <unknown>
  24: <unknown>
  25: g_cclosure_marshal_generic_va
  26: <unknown>
  27: g_signal_emit_valist
  28: g_signal_emit
  29: <unknown>
  30: <unknown>
  31: g_signal_emit_valist
  32: g_signal_emit
  33: <unknown>
  34: <unknown>
  35: g_signal_emit_valist
  36: g_signal_emit
  37: <unknown>
  38: g_cclosure_marshal_VOID__BOXEDv
  39: <unknown>
  40: g_signal_emit_valist
  41: g_signal_emit
  42: <unknown>
  43: <unknown>
  44: <unknown>
  45: <unknown>
  46: <unknown>
  47: <unknown>
  48: <unknown>
  49: <unknown>
  50: <unknown>
  51: <unknown>
  52: <unknown>
  53: g_closure_invoke
  54: <unknown>
  55: g_signal_emit_valist
  56: g_signal_emit
  57: <unknown>
  58: <unknown>
  59: g_main_context_dispatch
  60: <unknown>
  61: g_main_context_iteration
  62: g_application_run
  63: <O as gio::application::ApplicationExtManual>::run_with_args
             at ./Projects/telegrand/_build/cargo-home/registry/src/github.com-1ecc6299db9ec823/gio-0.14.8/src/application.rs:30:13
  64: <O as gio::application::ApplicationExtManual>::run
             at ./Projects/telegrand/_build/cargo-home/registry/src/github.com-1ecc6299db9ec823/gio-0.14.8/src/application.rs:23:9
  65: telegrand::main
             at ./Projects/telegrand/src/main.rs:78:5
  66: core::ops::function::FnOnce::call_once
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

I'll comment again if I have more info to share.

Reply:
Apparently this session is actually already closed in this case? That would make sense, but then I wonder if we're also receiving the right authentication state.

Reply:
Also because I'm getting other panics from `set_active_client_online()` when focusing the window, this should be handled a bit better, we just need to figure out how.

Reply:
Apparently we need to handle the `LoggingOut` state to fix this.

Reply:
Wow, thanks for fuzzing that. That was indeed not tested by me. I tried to fix that. In the course of this, I also simplified the code a lot. Would be cool if you could test it again.

Reply:
It still crashes, but for a different reason now.

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { domain: g-io-error-quark, code: 1, message: "Error opening file /var/home/melix/.var/app/com.github.melix99.telegrand.Devel/data/telegrand/db1641949426723/profile_photos/1234150945927047089.jpg: No such file or directory" }', src/session/avatar.rs:125:63
stack backtrace:
   0: rust_begin_unwind
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/std/src/panicking.rs:517:5
   1: core::panicking::panic_fmt
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/panicking.rs:100:14
   2: core::result::unwrap_failed
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/result.rs:1616:5
   3: core::result::Result<T,E>::unwrap
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/result.rs:1298:23
   4: telegrand::session::avatar::Avatar::load
             at ./Projects/telegrand/src/session/avatar.rs:125:31
   5: telegrand::session::avatar::Avatar::set_needed
             at ./Projects/telegrand/src/session/avatar.rs:185:13
   6: telegrand::session::components::avatar::Avatar::request_avatar_image
             at ./Projects/telegrand/src/session/components/avatar.rs:109:13
   7: telegrand::session::components::avatar::Avatar::set_item
             at ./Projects/telegrand/src/session/components/avatar.rs:122:9
   8: <telegrand::session::components::avatar::imp::Avatar as glib::subclass::object::ObjectImpl>::set_property
             at ./Projects/telegrand/src/session/components/avatar.rs:71:27
   9: glib::subclass::object::set_property
             at ./Projects/telegrand/_build/cargo-home/registry/src/github.com-1ecc6299db9ec823/glib-0.14.8/src/subclass/object.rs:99:5
  10: <unknown>
  11: g_object_setv
  12: g_object_set_property
  13: <unknown>
  14: g_closure_invoke
  15: <unknown>
  16: g_signal_emit_valist
  17: g_signal_emit
  18: <unknown>
  19: g_object_notify
  20: <T as glib::object::ObjectExt>::notify
             at ./Projects/telegrand/_build/cargo-home/registry/src/github.com-1ecc6299db9ec823/glib-0.14.8/src/object.rs:2177:13
  21: telegrand::session::Session::fetch_me::{{closure}}::{{closure}}
             at ./Projects/telegrand/src/session/mod.rs:564:25
  22: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80:19
  23: telegrand::utils::do_async::{{closure}}
             at ./Projects/telegrand/src/utils.rs:202:9
  24: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/future/mod.rs:80:19
  25: <futures_task::future_obj::LocalFutureObj<T> as core::future::future::Future>::poll
             at ./Projects/telegrand/_build/cargo-home/registry/src/github.com-1ecc6299db9ec823/futures-task-0.3.19/src/future_obj.rs:84:18
  26: <glib::main_context_futures::FutureWrapper as core::future::future::Future>::poll
             at ./Projects/telegrand/_build/cargo-home/registry/src/github.com-1ecc6299db9ec823/glib-0.14.8/src/main_context_futures.rs:31:44
  27: glib::main_context_futures::TaskSource::poll::{{closure}}
             at ./Projects/telegrand/_build/cargo-home/registry/src/github.com-1ecc6299db9ec823/glib-0.14.8/src/main_context_futures.rs:206:13
  28: glib::main_context::<impl glib::auto::main_context::MainContext>::with_thread_default
             at ./Projects/telegrand/_build/cargo-home/registry/src/github.com-1ecc6299db9ec823/glib-0.14.8/src/main_context.rs:125:9
  29: glib::main_context_futures::TaskSource::poll
             at ./Projects/telegrand/_build/cargo-home/registry/src/github.com-1ecc6299db9ec823/glib-0.14.8/src/main_context_futures.rs:200:9
  30: glib::main_context_futures::TaskSource::dispatch
             at ./Projects/telegrand/_build/cargo-home/registry/src/github.com-1ecc6299db9ec823/glib-0.14.8/src/main_context_futures.rs:66:34
  31: g_main_context_dispatch
  32: <unknown>
  33: g_main_context_iteration
  34: g_application_run
  35: <O as gio::application::ApplicationExtManual>::run_with_args
             at ./Projects/telegrand/_build/cargo-home/registry/src/github.com-1ecc6299db9ec823/gio-0.14.8/src/application.rs:30:13
  36: <O as gio::application::ApplicationExtManual>::run
             at ./Projects/telegrand/_build/cargo-home/registry/src/github.com-1ecc6299db9ec823/gio-0.14.8/src/application.rs:23:9
  37: telegrand::main
             at ./Projects/telegrand/src/main.rs:78:5
  38: core::ops::function::FnOnce::call_once
             at /rustc/f1edd0429582dd29cccacaf50fd134b05593bd9c/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

I'm still not sure why the file doesn't exist anymore, maybe it is being deleted by `teardown_database()` a bit too soon?

Reply:
Hmm. I have a suspicion. I think I delete the database directory too early. I think it's better to only remove them on application start. 
Guess this needs a few more days.

Reply:
I was thinking, shouldn't this be handled by tdlib? E.g. the `logOut` function states that `all local data will be destroyed`.

Reply:
If I see it correctly, `logOut` only deletes the `*sqlite*` files and the `*.binlog` files but does not touch the directory.

Reply:
Right. Then we should delete the directory only when we know that the `Session` is being destroyed. So maybe doing that in the `dispose()` function could be a better idea. Another option is doing that right after removing the `Session` from the stack.

Reply:
Can you test again? I have hope that it will work now. The code has also been extremely stripped.

Reply:
Unfortunately I still get the same panic.

Reply:
The problem only occurs when I terminate the session while the app is closed and then I reopen it. It doesn't happen when I terminate it while the app is open.

Reply:
No crash this time (as expected). I'll try to debug a bit more.

Reply:
I think we should delete the directory on `AuthorizationState::Closed `. Therefore, I would create a third variant for the `ClientInfo` enum called `LoggingOut`.

Reply:
Ok, I did that. Hopefully this fixes it.

Reply:
Yep, it seems to work now. Nice work! :)

Reply:
> Also because I'm getting other panics from `set_active_client_online()` when focusing the window, this should be handled a bit better, we just need to figure out how.

Still experiencing this?

Reply:
Nope, the client removes the `Session` immediately, so it's perfect now :)